src/structarray.jl

```diff
@@ -532,12 +576,22 @@ end
 
 # Unwrapper to recover the behaviour defined by parent style.
 @inline function Base.copyto!(dest::AbstractArray, bc::Broadcasted{StructArrayStyle{S, N}}) where {S, N}
-    return copyto!(dest, convert(Broadcasted{S}, bc))
+    bc′ = _use_default_bc(S()) ? convert(Broadcasted{S}, bc) : replace_structarray(bc)
+    return copyto!(dest, bc′)
 end
 
 @inline function Broadcast.materialize!(::StructArrayStyle{S}, dest, bc::Broadcasted) where {S}
-    return Broadcast.materialize!(S(), dest, bc)
+    bc′ = _use_default_bc(S()) ? bc : replace_structarray(bc)
+    return Broadcast.materialize!(S(), dest, bc′)
 end
 
 # for aliasing analysis during broadcast
+function Broadcast.broadcast_unalias(dest::StructArray, src::AbstractArray)
+    if dest === src || any(Base.Fix2(===, src), components(dest))
```

Comment:
Just to make sure I understand, this is just an optimization (somewhat unrelated to the PR) for operations like

```julia
A .= f.(A.x, A.y)
```

where `:x` and `:y` are the components of `A`? In that case, I'm happy to keep the change but maybe also that case should get a unit test.

Reply:
I just forgot that I have add a test for this. (This was added to avoid possible allocation on `A .= A .+ 1` on non-native structarray broadcast.)
Your example is much more intuitive.